Adjust handle anchor offset for collapsed text selection for Android

muhammadkamel opened the pull request:
The text selection anchor/handle in InputField was slightly misaligned, shifting slightly to the left. This PR fixes the handle positioning to ensure proper alignment with the text cursor.

Before:
![baloon_before](https://github.com/user-attachments/assets/99d69ea0-9559-4b79-8402-aa7657dc47b9)


After fix:
![baloon_after_fixed](https://github.com/user-attachments/assets/d89d6a99-f026-4e59-bb17-ad8950aea6f0)



## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

muhammadkamel committed:
Adjust handle anchor offset for collapsed text selection in MaterialTextSelectionControls

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
update: ancher offset

muhammadkamel commented:
@kenzieschmoll 
Appreciate your support to check this :)

Piinks commented:
@muhammadkamel code reviewers are assigned during weekly triage, please give us time to do so instead of pinging maintainers. They may not have any knowledge of this part of the codebase.
https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#who

muhammadkamel commented:
> @muhammadkamel code reviewers are assigned during weekly triage, please give us time to do so instead of pinging maintainers. They may not have any knowledge of this part of the codebase. https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#who

Hi @Piinks 

Got it!
Thanks for your support!

justinmc commented on packages/flutter/lib/src/material/text_selection.dart:
```diff
@@ -112,7 +112,7 @@ class MaterialTextSelectionControls extends TextSelectionControls {
   Offset getHandleAnchor(TextSelectionHandleType type, double textLineHeight) {
     return switch (type) {
-      TextSelectionHandleType.collapsed => const Offset(_kHandleSize / 2, -4),
+      TextSelectionHandleType.collapsed => const Offset(_kHandleSize / 2.18, -4.5),
```
Can you add a comment explaining how you came up with these values? Something like:

```
// These values were eyeballed to match a physical Pixel 9 Running Android 16.
```

justinmc reviewed (commented):
Thank you for fixing this! A one line fix but I think it makes it look a lot better.

Can you add a test? Probably something in text_selection_test.dart that shows the handle, gets the handle's Rect and the cursor's Rect, and then verifies the position of the handle in relation to the cursor.

I took a screenshot of my Pixel 9 Android 16 running the Google Keep app and zoomed in to confirm that the handle position matches what is shown in this PR 👍 .

muhammadkamel committed:
Update text_selection.dart

Updated the calculation of the anchor point for TextSelectionHandleType.collapsed to properly account for the cursor width.

Changed from:
Offset(_kHandleSize / 2, -4.6)
to:
Offset((_kHandleSize - cursorWidth) / 2, -4.6)

This ensures that the anchor aligns with the actual visual center of the cursor.
This change is necessary because the total handle size is 48x48 pixels, which is larger than the actual cursor width (2px), so centering must be adjusted accordingly.

muhammadkamel committed:
Add test for getHandleAnchor in MaterialTextSelectionControls

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel commented:
@justinmc 

Also, this is a new screenshot after applying this calculation:
![fix_ancher](https://github.com/user-attachments/assets/239269aa-5f79-4956-baf8-c45471bbf5b0)

muhammadkamel commented:
Hi @justinmc 

Appreciate your support to check the new updates

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

justinmc commented on packages/flutter/lib/src/material/text_selection.dart:
```diff
@@ -112,7 +112,7 @@ class MaterialTextSelectionControls extends TextSelectionControls {
   Offset getHandleAnchor(TextSelectionHandleType type, double textLineHeight) {
     return switch (type) {
-      TextSelectionHandleType.collapsed => const Offset(_kHandleSize / 2, -4),
+      TextSelectionHandleType.collapsed => const Offset((_kHandleSize - 2) / 2, -4.6),
```
I still think you should add a comment here explaining where these values come from. See https://github.com/flutter/flutter/pull/169715#discussion_r2127495258.

justinmc replied:
Ok based on your comments above like https://github.com/flutter/flutter/pull/169715#issuecomment-3009214569, it sounds like a comment like this would be good:

 > Eyeballed to match a physical Pixel 9 running Android 16, which horizontally centers the anchor 1 pixel below the cursor.

I can confirm that that's what my Pixel 9 seems to do.

justinmc commented on packages/flutter/test/material/text_selection_test.dart:
```diff
@@ -768,4 +768,24 @@ void main() {
   );
-}
+
+  group('getHandleAnchor', () {
+    test('returns correct anchor for collapsed handle', () {
+      const Offset expected = Offset((22.0 - 2.0) / 2, -4.6);
+      final Offset actual = materialTextSelectionControls.getHandleAnchor(TextSelectionHandleType.collapsed, 0);
+      expect(actual, equals(expected));
+    });
+
+    test('returns correct anchor for left handle', () {
+      const Offset expected = Offset(22.0, 0);
+      final Offset actual = materialTextSelectionControls.getHandleAnchor(TextSelectionHandleType.left, 0);
+      expect(actual, equals(expected));
+    });
+
+    test('returns correct anchor for right handle', () {
+      const Offset expected = Offset.zero;
+      final Offset actual = materialTextSelectionControls.getHandleAnchor(TextSelectionHandleType.right, 0);
+      expect(actual, equals(expected));
+    });
+  });
+}
```
I think these tests are not very useful since they just confirm the hardcoded values in the method. Could you add a test like I suggested in https://github.com/flutter/flutter/pull/169715#pullrequestreview-2898194726?

muhammadkamel commented:
@justinmc 
And it is look like this when I add 1px to the padding without change the offset values like in the PR commits...

![Screenshot 2025-06-26 at 8 22 48 PM](https://github.com/user-attachments/assets/cb7fffa3-9fe2-4578-a52c-afbdb021bc45)

justinmc reviewed (commented):
The things I see that this PR still needs before merge:

  * Add a comment explaining the value. I wrote out the text for you in my comment in this review.
  * Add a test along the lines of https://github.com/flutter/flutter/pull/169715#pullrequestreview-2898194726. Let me know if you need help with that.
  * Fix the failing checks.

justinmc commented:
@muhammadkamel Do you still have time to finish this PR? We definitely want to get this fixed! Thanks.

muhammadkamel commented:
I’ll be able to continue working on this PR in the next few days, insha'Allah.

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel commented:
Hi @justinmc,

I've added a border to the anchor container, which resolves the positioning issue of the anchor in the input field like in this image.

<img width="487" height="767" alt="Screenshot 2025-07-23 at 4 31 25 PM" src="https://github.com/user-attachments/assets/33eb42f5-322d-467b-bcaf-81466de464d4" />


Would you prefer keeping the border as a fix (with transparent color), or should we instead increase the padding of the anchor's parent by 1px on all sides (LTRB)?

justinmc commented:
Thanks for following up here! Doing it with padding seems to make more sense and be more semantic. I'm not sure exactly what that would look like in the code though, so I'd have to see it.

muhammadkamel committed:
Merge remote-tracking branch 'origin/master' into fix/text-selection-android-anchor

muhammadkamel committed:
feat: add cursorWidth parameter to getHandleAnchor method in text selection controls

muhammadkamel commented:
Hi @justinmc 

I have updated the behavior of the anchor (balloon) to be in the middle of the cursor width like the Android native behavior.

Like this image:

<img width="1080" height="2280" alt="Screenshot_1754282416" src="https://github.com/user-attachments/assets/55be3f7c-75fe-42d7-bc6d-835f0d6c40a0" />

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

justinmc reviewed (commented):
Sorry for the slow re-review here.

If I change the value of cursorWidth, does this actually change the width that's drawn, or just the value that's used in getHandleAnchor? Can you point me to where, if so?

justinmc commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -1254,6 +1268,12 @@ class SelectionOverlay {
   /// Called when the users start dragging the start selection handles.
   final ValueChanged<DragStartDetails>? onStartHandleDragStart;
 
+  /// Cursor width
+  /// This value is used for calculating the position of the text selection
+  /// handles.
+  ///
```
This still needs to be fixed.

justinmc commented on packages/flutter/lib/src/cupertino/desktop_text_selection.dart:
```diff
@@ -70,7 +70,11 @@ class CupertinoDesktopTextSelectionControls extends TextSelectionControls {
+  Offset getHandleAnchor(
+    TextSelectionHandleType type,
+    double textLineHeight, {
+    required double cursorWidth,
```
This could be a breaking change. Maybe make it not required and include the default 2.0 value here if it's not provided?

justinmc commented on packages/flutter/lib/src/material/selection_area.dart:
```diff
@@ -99,6 +100,9 @@ class SelectionArea extends StatefulWidget {
   /// {@macro flutter.widgets.ProxyWidget.child}
   final Widget child;
 
+  /// The width of the cursor.
+  final double cursorWidth;
```
I tried this out locally. It seems to have no effect on anything but collapsed cursors, so it never seems to do anything in non-editable text.

justinmc commented on packages/flutter/test/material/text_selection_test.dart:
```diff
@@ -776,4 +776,36 @@ void main() {
+        cursorWidth: 2.0,
+      );
+      expect(actual, equals(expected));
+    });
```
Could you add a test confirming the behavior if the cursor is wider than 2.0?

justinmc commented on packages/flutter/lib/src/material/text_selection.dart:
```diff
@@ -110,9 +110,13 @@ class MaterialTextSelectionControls extends TextSelectionControls {
+  }) {
     return switch (type) {
-      TextSelectionHandleType.collapsed => const Offset(_kHandleSize / 2, -4),
+      TextSelectionHandleType.collapsed => -Offset((cursorWidth - _kHandleSize) / 2, 5),
```
So the cursor width extends to the right of the collapsed selection location (this is existing behavior). Given that it's not centered, we do have to consider the cursor's width when placing the handle.

This makes me wonder, though, should we be centering the cursor and keeping the handle in the same place? What are native platforms doing here?

Here's what it looks like with an extra wide cursor:

| master | fix/text-selection-android-anchor |
| --- | --- |
| <img width="100" height="93" alt="Image" src="https://github.com/user-attachments/assets/35a94afd-f4d6-45f6-a5fd-5f48096bd1a6" /> | <img width="110" height="154" alt="Image" src="https://github.com/user-attachments/assets/b5bfe536-03b2-4d8d-afb2-06f086636f6e" /> |

muhammadkamel replied:
I have tried a native example in Kotlin by changing the cursor width and the handle anchor were in the middle of the cursor.

Like:

If the cursor width value is 50px, the handle anchor balloon were in the middle of it.

justinmc replied:
Thank you for checking! Sounds like we should stick with the positioning you already have then 👍 .

muhammadkamel replied:
You're welcome 🙂
What would you like me to do next?

justinmc replied:
Can you fix the failing checks?

justinmc reviewed (commented):
Would it be reasonable to reference that 2.0 pixel width default value in one place? Maybe on TextSelectionOverlay or somewhere add `static const double defaultCursorWidth = 2.0;` and reference that everywhere you have 2.0 hardcoded now.

muhammadkamel commented:
> @muhammadkamel Can you take a look at this thread? https://github.com/flutter/flutter/pull/173511/files#r2277776618 It seems like this PR is lacking a vertical position solution as seen in that PR.

@justinmc I believe the issue only affects the text selection cursor and its anchor (balloon), which I fixed in this PR.  
I've handled the `collapsed` handle case, so `right` and `left` types should now behave as expected.

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

justinmc reviewed (commented):
I left a comment about some math that needs fixing. Also heads up that you have failures on this PR.

muhammadkamel committed:
Merge branch 'flutter:master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'flutter:master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

justinmc commented:
Looks like there are still failures @muhammadkamel.

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

Renzo-Olivares commented:
@muhammadkamel Just a heads up this PR is still running into failures. Please feel free to reach out flutter hackers discord for help on these.

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor
